### PR TITLE
feat(api): add getAddresses api implementation

### PR DESCRIPTION
**Pull request by edgarkhanzadian**
Move implementation of getAddresses api to monorepo. It is going to be exported as a `useGetAddresses` hook that is going to be used in `src/app/pages/rpc-get-addresses/use-request-accounts.ts` file in extension.

To test this yourself, run `yarn link` in `mono/packages/crypto` and then run `yarn link @leather-wallet/crypto` in extension repo. After that, replace your code in `use-request-accounts.ts` file with this code:

```js
import { useGetAddresses as useNewGetAddresses } from '@leather-wallet/crypto/src/getAddresses';
import { useAtomValue } from 'jotai';

import { logger } from '@shared/logger';
import { makeRpcSuccessResponse } from '@shared/rpc/rpc-methods';
import { closeWindow } from '@shared/utils';

import { useAnalytics } from '@app/common/hooks/analytics/use-analytics';
import { useRpcRequestParams } from '@app/common/rpc-helpers';
import { hasSwitchedAccountsState, useCurrentAccountIndex } from '@app/store/accounts/account';
import { useNativeSegwitAccountBuilder } from '@app/store/accounts/blockchain/bitcoin/native-segwit-account.hooks';
import { useTaprootAccount } from '@app/store/accounts/blockchain/bitcoin/taproot-account.hooks';
import { useStacksAccounts } from '@app/store/accounts/blockchain/stacks/stacks-account.hooks';
import { useAppPermissions } from '@app/store/app-permissions/app-permissions.slice';
import { useHasLedgerKeys } from '@app/store/ledger/ledger.selectors';
import { useCurrentNetwork } from '@app/store/networks/networks.selectors';
import { useSignatureRequestAccountIndex } from '@app/store/signatures/requests.hooks';
import { useCurrentKeyDetails } from '@app/store/software-keys/software-key.selectors';
import { useTransactionRequestState } from '@app/store/transactions/requests.hooks';

export function useGetAddresses() {
  const analytics = useAnalytics();

  const currentAccountIndex = useCurrentAccountIndex();
  // ⚠️ to refactor, we should not just continually add new conditionals here
  const hasSwitchedAccounts = useAtomValue(hasSwitchedAccountsState);
  const stacksAccounts = useStacksAccounts();
  const segwitAccount = useNativeSegwitAccountBuilder()(currentAccountIndex);
  const taprootAccount = useTaprootAccount(currentAccountIndex);
  const signatureIndex = useSignatureRequestAccountIndex();
  const currentNetwork = useCurrentNetwork();
  const wallet = useCurrentKeyDetails();
  const isLedger = useHasLedgerKeys();
  const transactionPayload = useTransactionRequestState();

  const { stacksAddressResponse, nativeSegwitAddressResponse, taprootAddressResponse } =
    useNewGetAddresses({
      signatureIndex,
      currentAccountIndex,
      currentNetwork,
      hasSwitchedAccounts,
      segwitAccount,
      stacksAccounts,
      taprootAccount,
      wallet,
      hasLedgerKeys: isLedger,
      transactionPayload,
    });

  const permissions = useAppPermissions();
  const { tabId, origin, requestId } = useRpcRequestParams();

  return {
    origin,
    onUserApproveGetAddresses() {
      if (!tabId || !origin) {
        logger.error('Cannot give app accounts: missing tabId, origin');
        return;
      }
      void analytics.track('user_approved_get_addresses', { origin });
      permissions.hasRequestedAccounts(origin);
      chrome.tabs.sendMessage(
        tabId,
        makeRpcSuccessResponse('getAddresses', {
          id: requestId,
          result: {
            addresses: [
              nativeSegwitAddressResponse,
              taprootAddressResponse,
              // casting until btckit has better extend support
              stacksAddressResponse as any,
            ],
          },
        })
      );
      closeWindow();
    },
  };
}

```

### CODE REVIEW

File: packages/crypto/src/utils/useWalletType.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
+enum WalletType {
+  Ledger = 'ledger',
+  Software = 'software',
+}
+
+function isLedgerWallet(walletType: WalletType) {
```
Comment (kyranjamie) [inline packages/crypto/src/utils/useWalletType.ts:6]: File uses camelcase, let's stick to kebab case

File: packages/crypto/src/getAddresses.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  signatureIndex,
+  currentAccountIndex,
+  currentNetwork,
+  hasLedgerKeys,
+  hasSwitchedAccounts,
+  segwitAccount,
+  stacksAccounts,
+  taprootAccount,
+  transactionPayload,
+  wallet,
```
Comment (kyranjamie) [inline packages/crypto/src/getAddresses.ts:23]: This implementation means that the `@leather/crypto` package is "aware" of all the chains we support.

Say we had more chains in future, we'd have to come to this function an expand it with more object properties. What if this factory hook instead was responsible only for formatting, and delivering the response?

Then the API might look something like:

```ts
interface UseGetAddressesProps {
  onApproveGetAddresses(accounts: accounts[]): Promise<void>;
  accounts[]: SomeMoreGenericTypeForAnyChain;
}
function useGetAddresses({ onApproveGetAddresses, accounts }: UseGetAddressesProps) {}

// extension
useGetAddreseses({ 
  // type checking ensures these conform to standard
  accounts,
  onApproveGetAddresses(accounts) {
    // specific extension
    chrome.tabs.sendMessage(tabId, accounts);
  }
})

// RN
useGetAddreseses({ 
  accounts,
  onApproveGetAddresses(accounts) {
    // specific RN api
    somethingElseThatCommunitcatesWithWebView();
  }
})
```

Here, the responsibility is more scoped to just formatting, checking, and passing on the response.

File: packages/crypto/src/getAddresses.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+}: {
+  signatureIndex: number | undefined;
+  currentAccountIndex: number;
+  currentNetwork: NetworkConfiguration;
+  hasLedgerKeys: boolean;
+  hasSwitchedAccounts: boolean;
+  segwitAccount: BitcoinAccount | undefined;
+  stacksAccounts: StacksAccount[];
+  taprootAccount: BitcoinAccount | undefined;
+  transactionPayload: TransactionPayload | null;
```
Comment (kyranjamie) [inline packages/crypto/src/getAddresses.ts:33]: This relates to legacy stacks auth responses we want to remove.

File: packages/crypto/src/getAddresses.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  wallet: KeyConfig | undefined;
```
Comment (kyranjamie) [inline packages/crypto/src/getAddresses.ts:34]: Confusing naming as it's specific to stacks. Only `@leather/stacks` should know of these details

**Review by kyranjamie — COMMENTED** [review]
Thanks for taking this on @edgarkhanzadian. Maybe we can just on a call to talk through this later today/next week.

I think we can simplify the implementation slightly, divide the responsibilities, and have the API part of this focused more on formatting/serving the data than

**Comment by markmhendrickson** [comment]
To be linked to the individual package-creation issues that come out of https://github.com/leather-wallet/extension/issues/4297

**Comment by markmhendrickson** [comment]
To finish PR but without getAddresses part

**Comment by markmhendrickson** [comment]
@edgarkhanzadian should this undergo review again?

**Comment by pete-watters** [comment]
@edgarkhanzadian - can we close this or else finish the work?

**Comment by kyranjamie** [comment]
@pete-watters this is on my plate

**Pull request closed without merging**